sympy/core/function.py

```diff
@@ -1894,34 +1911,73 @@ class Lambda(Expr):
-        from sympy.sets.sets import FiniteSet
-        v = list(variables) if iterable(variables) else [variables]
-        for i in v:
-            if not getattr(i, 'is_symbol', False):
-                raise TypeError('variable is not a symbol: %s' % i)
-        if len(v) != len(set(v)):
-            x = [i for i in v if v.count(i) > 1][0]
-            raise SyntaxError("duplicate argument '%s' in Lambda args" % x)
-        if len(v) == 1 and v[0] == expr:
+    def __new__(cls, signature, expr):
+        if iterable(signature) and not isinstance(signature, (tuple, Tuple)):
+            SymPyDeprecationWarning(
+                feature="non tuple iterable of argument symbols to Lambda",
+                useinstead="tuple of argument symbols",
+                issue=17474,
+                deprecated_since_version="1.5").warn()
+            signature = tuple(signature)
+        sig = signature if iterable(signature) else (signature,)
+        sig = sympify(sig)
+        cls._check_signature(sig)
+
+        if len(sig) == 1 and sig[0] == expr:
             return S.IdentityFunction
 
-        obj = Expr.__new__(cls, Tuple(*v), sympify(expr))
-        obj.nargs = FiniteSet(len(v))
-        return obj
+        return Expr.__new__(cls, sig, sympify(expr))
+
+    @classmethod
+    def _check_signature(cls, sig):
+        syms = set()
+
+        def rcheck(args):
+            for a in args:
+                if a.is_symbol:
+                    if a in syms:
+                        raise BadSignatureError("Duplicate symbol %s" % a)
+                    syms.add(a)
+                elif isinstance(a, Tuple):
+                    rcheck(a)
+                else:
+                    raise BadSignatureError("Lambda signature should be only tuples"
+                        " and symbols, not %s" % a)
+
+        if not isinstance(sig, Tuple):
+            raise BadSignatureError("Lambda signature should be a tuple not %s" % sig)
+        # Recurse through the signature:
+        rcheck(sig)
 
     @property
-    def variables(self):
-        """The variables used in the internal representation of the function"""
+    def signature(self):
+        """The expected form of the arguments to be unpacked into variables"""
         return self._args[0]
 
-    bound_symbols = variables
-
     @property
     def expr(self):
         """The return value of the function"""
         return self._args[1]
 
+    @property
+    def variables(self):
+        """The variables used in the internal representation of the function"""
+        def _variables(args):
+            if isinstance(args, Tuple):
+                for arg in args:
+                    for a in _variables(arg):
+                        yield a
+            else:
+                yield args
+        return tuple(_variables(self.signature))
+
+    @property
+    def nargs(self):
+        from sympy.sets.sets import FiniteSet
+        return FiniteSet(len(self.signature))
```

Comment:
Will this always be a set containing a single int? Could it be just the int?

Reply:
Maybe. I'm just going with what already happens for other functions here though:
```julia
In [19]: sin.nargs                                                                                                                
Out[19]: {1}

In [20]: atan2.nargs                                                                                                              
Out[20]: {2}

In [21]: log.nargs                                                                                                                
Out[21]: {1, 2}

In [22]: type(_)                                                                                                                  
Out[22]: sympy.sets.sets.FiniteSet
```
I'm also not sure why this needs to be a symbolic `FiniteSet` rather than a normal Python `set/frozenset`. I don't see how this is used in a symbolic context... actually some functions have `nargs == Naturals0`.

Reply:
I see, it has to be a set as there are functions with variable number of arguments.